### PR TITLE
 Build containerd and runc for specific arch with xx

### PR DESCRIPTION
**Pull request by manugupt1**

### CODE REVIEW

**Comment by manugupt1** [comment]
cc @AkihiroSuda is something like this acceptable.
We build only for the TARGETARCH.
Just doing that saves more than 2 minutes per build.

File: Dockerfile [hunk 1]
```diff
@@ -60,32 +60,30 @@ RUN dpkg --add-architecture arm64 && \
   apt-get install -y crossbuild-essential-amd64 crossbuild-essential-arm64 git libbtrfs-dev:amd64 libbtrfs-dev:arm64 libseccomp-dev:amd64 libseccomp-dev:arm64
 
 FROM build-base-debian AS build-containerd
+ARG TARGETARCH
 ARG CONTAINERD_VERSION
 RUN git clone https://github.com/containerd/containerd.git /go/src/github.com/containerd/containerd
+RUN TARGETARCH=${TARGETARCH:-amd64}
```
Comment (AkihiroSuda) [inline Dockerfile:66]: wrong

File: Dockerfile [hunk 2]
```diff
@@ -60,32 +60,30 @@ RUN dpkg --add-architecture arm64 && \
...
 WORKDIR /go/src/github.com/opencontainers/runc
 RUN git checkout ${RUNC_VERSION} && \
   mkdir -p /out
 ENV CGO_ENABLED=1
-RUN GOARCH=amd64 CC=x86_64-linux-gnu-gcc make static && \
-  cp -a runc /out/runc.amd64
-RUN GOARCH=arm64 CC=aarch64-linux-gnu-gcc make static && \
-  cp -a runc /out/runc.arm64
+RUN TARGETARCH=${TARGETARCH:-amd64} GOARCH=${TARGETARCH} CC=`[ "${TARGETARCH}" = "arm64" ] && echo aarch64-linux-gnu-gcc || echo x86_64-linux-gnu-gcc` make static && \
+  cp -a runc /out/runc.${TARGETARCH}
```
Comment (AkihiroSuda) [inline Dockerfile:86]: Shouldn't we use https://github.com/tonistiigi/xx

Comment (manugupt1) [reply Dockerfile:86]: I had no idea about this; let me check this out.
Thanks! Will get back.

File: Dockerfile (hunk 1 above)
Comment (manugupt1) [reply Dockerfile:66]: Hmm! I did not understand this at a lot of places here; we default to amd64 if not present. 
What do you mean?

File: Dockerfile [hunk 3]
```diff
@@ -51,53 +51,57 @@ ARG UBUNTU_VERSION=22.04
...
 ARG GOTESTSUM_VERSION=v1.9.0
 
+FROM --platform=$BUILDPLATFORM tonistiigi/xx:1.2.1 AS xx
+
+
 FROM --platform=$BUILDPLATFORM golang:${GO_VERSION}-bullseye AS build-base-debian
+COPY --from=xx / /
+ENV DEBIAN_FRONTEND=noninteractive
+RUN apt-get update && \
+  apt-get install -y git pkg-config crossbuild-essential-amd64 crossbuild-essential-arm64
```
Comment (AkihiroSuda) [inline Dockerfile:61]: `crossbuild-essential-*` does not need to be specified here

File: Dockerfile [hunk 4]
```diff
@@ -51,53 +51,57 @@ ARG UBUNTU_VERSION=22.04
...
 WORKDIR /go/src/github.com/opencontainers/runc
 RUN git checkout ${RUNC_VERSION} && \
   mkdir -p /out
 ENV CGO_ENABLED=1
-RUN GOARCH=amd64 CC=x86_64-linux-gnu-gcc make static && \
-  cp -a runc /out/runc.amd64
-RUN GOARCH=arm64 CC=aarch64-linux-gnu-gcc make static && \
-  cp -a runc /out/runc.arm64
+RUN xx-go --wrap
+RUN GOARCH=${TARGETARCH} CC=$(xx-info)-gcc make static && \
```
Comment (AkihiroSuda) [inline Dockerfile:91]: `RUN GO=xx-go make ...` should work?

File: Dockerfile [hunk 5]
```diff
@@ -51,53 +51,56 @@ ARG UBUNTU_VERSION=22.04
...
-  mkdir -p /out/amd64 /out/arm64
+  mkdir -p /out/${TARGETARCH}
 ENV CGO_ENABLED=1
-RUN GOARCH=amd64 CC=x86_64-linux-gnu-gcc make static && \
-  cp -a bypass4netns bypass4netnsd /out/amd64
-RUN GOARCH=arm64 CC=aarch64-linux-gnu-gcc make static && \
-  cp -a bypass4netns bypass4netnsd /out/arm64
+RUN xx-go --wrap
+RUN GO=xx-go GOARCH=${TARGETARCH} CC=$(xx-info)-gcc make static && \
+  xx-verify --static bypass4netns && cp -a bypass4netns bypass4netnsd /out/${TARGETARCH}
```
Comment (AkihiroSuda) [inline Dockerfile:103]: When GO=xx-go is set, no need to run --wrap and specify GOARCH, CC, etc.

Comment (manugupt1) [reply Dockerfile:103]: thanks! sorry! misunderstood earlier. Thanks for helping with this PR throughout

File: Dockerfile [hunk 6]
```diff
@@ -51,53 +51,56 @@ ARG UBUNTU_VERSION=22.04
...
 WORKDIR /go/src/github.com/rootless-containers/bypass4netns
 RUN git checkout ${BYPASS4NETNS_VERSION} && \
-  mkdir -p /out/amd64 /out/arm64
+  mkdir -p /out/${TARGETARCH}
 ENV CGO_ENABLED=1
-RUN GOARCH=amd64 CC=x86_64-linux-gnu-gcc make static && \
-  cp -a bypass4netns bypass4netnsd /out/amd64
-RUN GOARCH=arm64 CC=aarch64-linux-gnu-gcc make static && \
-  cp -a bypass4netns bypass4netnsd /out/arm64
+RUN xx-go --wrap
```
Comment (manugupt1) [inline Dockerfile:101]: Remove

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**